### PR TITLE
feat: add live training dashboard page with recorded CartPole curves

### LINKED ISSUE #264: Live training dashboard in the web demo
## Summary
Extend the web demo with a live training dashboard. Roadmap Phase 4 item (the only remaining unchecked Phase 4 box). Current demos at https://rjwalters.github.io/thrust/ are inference-only (pre-trained policies).

## Motivation
A "watch it learn" view — reward/loss curves updating during training — is the highest-visibility showcase of the train-in-Rust story and complements the existing inference demos.

## Scope / pointers
- Frontend: `web/src/pages/` (React + Vite; existing pages: CartPole, Snake, Pong, bandit).
- Examples already emit learning curves via the `CURVE_CSV` env var — a natural data source, or stream metrics from a WASM training loop.
- Decide: replay a recorded training run (simple) vs. live in-browser training via the WASM bindings (`src/wasm.rs`, harder).

## Acceptance criteria
- [ ] A dashboard page showing reward/loss over training steps for at least one env.
- [ ] Wired into the site nav (`web/src/App.tsx`) and deploys via the existing Pages workflow.
- [ ] Document the data source (recorded CSV vs. live WASM training).

## Notes
Recommend starting with replay-of-recorded-run to ship value fast, then consider live in-browser training as a follow-up.

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary
Adds a live training dashboard to the web demo (Roadmap Phase 4). A new `/dashboard` page overlays reward-vs-steps learning curves for three RL algorithms — A2C, DQN and PPO — on CartPole-v1, rendered with recharts. This is the "watch it learn" showcase that complements the existing inference-only demos.

Approach is replay-of-recorded-run (per the curator guidance): the curves are real data recorded locally by the repository's training examples and committed as static CSVs under `web/public/`. No training runs in the browser.

## Changes
- **`web/src/pages/TrainingDashboardPage.tsx`** (new): fetches the three CSVs via `import.meta.env.BASE_URL` (same pattern as `CartPolePage`), parses the shared `env_steps,mean_episode_reward` schema, merges the series by env step, and renders a multi-series `LineChart` with labelled axes, legend, tooltip, a documented data-source note, and a prose section explaining each algorithm.
- **`web/src/App.tsx`**: add `<Route path="/dashboard" element={<TrainingDashboardPage />} />`.
- **`web/src/pages/Home.tsx`**: add a "Training Dashboard" card in the existing grid.
- **`web/package.json` / `web/pnpm-lock.yaml`**: add `recharts`.
- **`web/public/cartpole_curves_{a2c,dqn,ppo}.csv`** (new): recorded learning curves.
- **`web/tests/site-loads.spec.ts`**: add a "training dashboard page loads" smoke test (runs against the local dev server, asserts the `<h1>` contains "Training" and the recharts chart renders with no page errors).

## Data source
Generated locally by running the three CartPole training examples with `CURVE_CSV` set, all sharing seed 0 and a 60k env-step budget so the curves overlay directly:
```
TOTAL_TIMESTEPS=60000 CURVE_CSV=web/public/cartpole_curves_a2c.csv cargo run --release --features training --example train_cartpole_a2c
TOTAL_TIMESTEPS=60000 CURVE_CSV=web/public/cartpole_curves_dqn.csv cargo run --release --features training --example train_cartpole_dqn
TOTAL_TIMESTEPS=60000 CURVE_CSV=web/public/cartpole_curves_ppo.csv cargo run --release --features training --example train_cartpole_modern
```
This is real recorded data (not fabricated). The data source is also documented on the page itself. No Rust files or CI workflows were modified.

## Acceptance Criteria Verification

| Criterion | Status | Verification |
|-----------|--------|--------------|
| `/dashboard` page reachable from nav (Home card + App.tsx route) | ✅ | Route added; Home card links to it; `curl /thrust/dashboard` → 200 |
| Reward curve chart for ≥1 env, env_steps x-axis / mean_episode_reward y-axis | ✅ | recharts LineChart with labelled axes |
| ≥1 algorithm series (A2C+DQN+PPO stretch goal) | ✅ | All three overlaid |
| Data source documented on the page | ✅ | Note: "replay of recorded training runs — CartPole-v1, seed 0, 60k env steps" |
| Wired into App.tsx + linked from Home.tsx | ✅ | Both edited |
| `pnpm build` passes, no TS warnings | ✅ | `tsc -b && vite build` clean |
| Loads cleanly in Playwright smoke suite | ✅ | New test passes |
| Deploy picks up page automatically, no workflow change | ✅ | Static assets under web/public/, no CI edits |

## Test Plan
- `pnpm build` — passes (tsc + vite, no errors/warnings).
- `pnpm lint` on the four touched source files — clean (the 6 repo-wide errors are pre-existing in untouched files: `vite-env.d.ts`, `debug-games.spec.ts`, `production.spec.ts`).
- Playwright "training dashboard page loads" — passes: title is "Training Dashboard", recharts surface visible, zero page errors.

Note: Playwright's bundled Chromium could not be downloaded in this environment, so the smoke test was verified against locally installed Google Chrome (`channel: chrome`); the test itself is unchanged and standard.

Closes #264

### CODE REVIEW

**Comment by rjwalters** [comment]
✅ **Approved!**

Thorough review of the training dashboard. This is clean, well-structured work that meets all of #264's acceptance criteria.

## Verification performed
- **Build**: `pnpm install && pnpm build` in the issue-264 worktree — `tsc -b && vite build` clean, no TS errors/warnings. The three CSVs are correctly emitted to `dist/` (`cartpole_curves_{a2c,dqn,ppo}.csv`), so the existing Pages workflow picks them up with no CI change.
- **Lint (touched files)**: `eslint` on `TrainingDashboardPage.tsx`, `App.tsx`, `Home.tsx`, `site-loads.spec.ts` — clean. Confirmed the 6 repo-wide errors are pre-existing in untouched files.
- **CI**: `gh pr checks 284` — all 9 checks pass. Merge state CLEAN.
- **Lockfile**: `pnpm install` reports "Lockfile is up to date"; `recharts ^3.9.2` present in both `package.json` and `pnpm-lock.yaml`. No drift.
- **Stray files**: working tree clean; no throwaway Playwright/Chrome config committed (`web/playwright.config.ts` is the pre-existing standard config, untouched by this PR).

## Correctness — CSV parse/merge (the tricky part)
The three curves are on **disjoint env_step grids** (A2C at 1600-step intervals, DQN at 2000, PPO at 4096), so merging by exact step produces a sparse table where almost every `CurvePoint` holds a value for only one algorithm. This is handled correctly:
- `parseCurveCsv` drops the header, tolerates blank/malformed rows, and filters non-finite values.
- The `Map<step, CurvePoint>` merge plus `[...values()].sort()` yields a step-ordered dataset.
- Each `<Line>` uses `connectNulls`, so every series draws a continuous line through its own points across the interleaved gaps — the "missing points" edge case resolves cleanly. Coincident steps (if any) would also merge correctly into a single point.

## Data plausibility (not fabricated)
The recorded curves read as genuine runs, and the noise signatures match the prose:
- **A2C**: starts at 0, noisy, dips into single digits mid-run, recovers to ~32 — the high-variance instability A2C is known for on CartPole.
- **DQN**: climbs to ~102, dips to ~75 (a forgetting episode), recovers to ~159.
- **PPO**: smooth near-monotonic rise to ~229 — the stable performer, consistent with it being the inference-demo algorithm.
All `env_steps` are strictly monotonic. PPO stopping at 57344 (last 4096-rollout boundary before 60k) is expected, not a truncation bug.

## React/TS quality
- `import.meta.env.BASE_URL` fetch pattern matches `CartPolePage`; `Footer` reused; loading/error states handled; `cancelled` cleanup flag prevents setState-after-unmount. Labelled axes, legend, tooltip, and an on-page data-source note are all present.
- Route + Home card wiring is correct and consistent with the other cards.

## Minor (non-blocking)
The dashboard smoke test targets `LOCAL_URL` (`localhost:5173`) while the other tests hit the production `BASE_URL`, since the page isn't deployed until this merges. The divergence is documented in a comment and is the right call for now; it can be switched to the production URL in a follow-up once deployed.

Great work — approving.
